modules/ballerina-http/src/main/ballerina/ballerina/net/http/resiliency/circuit-breaker.bal

```diff
@@ -0,0 +1,368 @@
+package ballerina.net.http.resiliency;
+
+import ballerina.net.http;
+
+enum CircuitState {
+    OPEN, CLOSE, HALF_OPEN
+}
+
+struct CircuitHealth {
+    float requestCount;
+    float errorCount;
+    Time lastErrorTime;
+}
+
+public connector CircuitBreaker (http:HttpClient httpClient, float failurePercentageThreshold, int resetTimeout) {
+
+    endpoint<http:HttpClient> httpEP {
+        httpClient;
+    }
+
+    CircuitHealth circuitHealth = {};
+    // TODO: once enum init inside a struct is do-able, move this inside circuitHealth (issue #4340)
+    CircuitState currentState = CircuitState.CLOSE;
+
+    action post (string path, http:Request req) (http:Response, http:HttpConnectorError) {
+        currentState = updateCircuitState(circuitHealth, currentState, failurePercentageThreshold, resetTimeout);
+        http:Response res;
+        http:HttpConnectorError e;
+
+        if (currentState == CircuitState.CLOSE) {
+            res, e = httpEP.post(path, req);
+            circuitHealth.requestCount = circuitHealth.requestCount + 1;
+
+            if (e != null) {
+                circuitHealth.updateHealth();
+                http:Response errorResponse = createErrorResponse(e.msg, e.statusCode);
+                return errorResponse, e;
```

Comment:
Shall we move the below code block to a private method, as it is duplicated in multiple places?
`circuitHealth.requestCount = circuitHealth.requestCount + 1;

            if (e != null) {
                circuitHealth.updateHealth();
                http:Response errorResponse = createErrorResponse(e.msg, e.statusCode);
                return errorResponse, e;
            }`